akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.io
+
+import java.io.InputStream
+
+import akka.annotation.InternalApi
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.{
+  AbruptStageTerminationException,
+  Attributes,
+  IOOperationIncompleteException,
+  IOResult,
+  Outlet,
+  SourceShape
+}
+import akka.stream.stage.{ GraphStageLogic, GraphStageLogicWithLogging, GraphStageWithMaterializedValue, OutHandler }
+import akka.util.ByteString
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class InputStreamSource(factory: () => InputStream, chunkSize: Int)
+    extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+
+  require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+
+  private val out: Outlet[ByteString] = Outlet("InputStreamSource.out")
+
+  override def shape: SourceShape[ByteString] = SourceShape(out)
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.inputStreamSource
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val mat = Promise[IOResult]
+    val logic = new GraphStageLogicWithLogging(shape) with OutHandler {
+      private val buffer = new Array[Byte](chunkSize)
+      private var readBytesTotal = 0L
+      private var inputStream: InputStream = _
+      private def isClosed = mat.isCompleted
+
+      override def preStart(): Unit = {
+        try {
+          inputStream = factory()
+        } catch {
+          case NonFatal(t) =>
+            mat.failure(new IOOperationIncompleteException(0, t))
+            failStage(t)
+        }
+      }
+
+      override def onPull(): Unit = if (isAvailable(out)) {
+        try {
+          val readBytes = inputStream.read(buffer)
+          readBytes match {
+            case -1 =>
+              closeStage()
+            case _ =>
+              readBytesTotal += readBytes
+              push(out, ByteString.fromArray(buffer, 0, readBytes))
+          }
```

Comment:
Small style thing, could be a matter of taste, I think I would have matched directly on the return and capture the non `-1` value in a `readBytes` name instead of `_` and using the value from the outer scope.